File: tests/source/write_logs.py
"""Write some logs."""

import sys
import time
import os

from mdk import start
mdk = start()

start = time.time()

def got_logs(log_events):
    category = sys.argv[1]
    results = set()
    for event in log_events.result:
        print(event.category, event.text)
        if event.category == category:
            results.add((event.text, event.level))
    expected = set(("hello {} {}".format(category, level), level.upper())
                   for level in ["critical", "error", "warn", "info", "debug"])
    if expected == results:
        print("Got expected messages!")
        mdk.stop()
        # XXX working around Quark/MDK issue where Python runtime isn't exiting:
        os._exit(0)
    else:
        print("No full results yet, will try again. Specifically got: " + repr(results))
        read_logs()

def got_error(error):
    print("ERROR: " + error.toString())
    mdk.stop()
    os._exit(1)

def read_logs():
    if time.time() - start > 60:
        print("Took more than 60 seconds, giving up.")
        os._exit(1)
    tracer = mdk._tracer
    print(tracer.url)
    print(tracer.queryURL)
    now = int(start * 1000)
    tracer.query(now - 60000, now + 120000).andEither(got_logs, got_error)

def main():
    session = mdk.session()
    category = sys.argv[1]

    for level in ["critical", "error", "warn", "info", "debug"]:
        f = getattr(session, level)
        f(category, "hello {} {}".format(category, level))
    read_logs()

if __name__ == '__main__':
    main()

File: tests/source/write_logs.rb
require 'mdk'
$mdk = ::Quark::Mdk.start()

def main()
  start = Time.now
  session = $mdk.session()
  session.trace("DEBUG")
  messages = []
  category = ARGV[0]
  sent_messages = ["hello critical " + category,
		   "hello debug " + category,
		   "hello error " + category,
		   "hello info " + category,
		   "hello warn " + category]

  $mdk._tracer.subscribe(
    Proc.new {|event|
      if Time.now - start > 60 then
        puts "Timeout"
        exit! 1
      end
      puts "got message #{event}"
      if (event.category == category) then
        messages.push(event.text)
      end
      if (messages.length == 5) then
        $mdk.stop
        messages.sort!
        if (messages != sent_messages) then
          puts "unexpected responses:"
          puts messages
          exit! 1
        else
          puts "got all messages"
          exit! 0
        end
      end
    })

  session.critical(category, sent_messages[0])
  session.debug(category, sent_messages[1])
  session.error(category, sent_messages[2])
  session.info(category, sent_messages[3])
  session.warn(category, sent_messages[4])
end

main()
